src/runtime_src/core/pcie/driver/linux/xocl/subdev/accel_deadlock_detector.c

```diff
@@ -0,0 +1,288 @@
+    struct device		*dev;
+    uint64_t		start_paddr;
+    uint64_t		range;
+    struct mutex 		lock;
+    struct debug_ip_data	data;
+};
+
+static ssize_t name_show(struct device *dev,
+        struct device_attribute *attr, char *buf)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector = platform_get_drvdata(to_platform_device(dev));
+    return sprintf(buf, "accel_deadlock_%llu\n",accel_deadlock_detector->data.m_base_address);
+}
+
+static DEVICE_ATTR_RO(name);
+
+static ssize_t status_show(struct device *dev,
+        struct device_attribute *attr, char *buf)
+{
+    uint64_t result = 0;
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector = platform_get_drvdata(to_platform_device(dev));
+    mutex_lock(&accel_deadlock_detector->lock);
+    result = XOCL_READ_REG32(accel_deadlock_detector->base + 0x0);
+    mutex_unlock(&accel_deadlock_detector->lock);
+    return sprintf(buf, "%llu\n",result);
+}
+
+static DEVICE_ATTR_RO(status);
+
+static struct attribute *accel_deadlock_detector_attrs[] = {
+    &dev_attr_name.attr,
+    &dev_attr_status.attr,
+    NULL,
+};
+
+static struct attribute_group accel_deadlock_detector_attr_group = {
+    .attrs = accel_deadlock_detector_attrs,
+};
+
+static int accel_deadlock_detector_remove(struct platform_device *pdev)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector;
+    void *hdl;
+
+    accel_deadlock_detector = platform_get_drvdata(pdev);
+    if (!accel_deadlock_detector) {
+        xocl_err(&pdev->dev, "driver data is NULL");
+        return -EINVAL;
+    }
+
+    sysfs_remove_group(&pdev->dev.kobj, &accel_deadlock_detector_attr_group);
+
+    xocl_drvinst_release(accel_deadlock_detector, &hdl);
+
+    if (accel_deadlock_detector->base)
+        iounmap(accel_deadlock_detector->base);
+
+    platform_set_drvdata(pdev, NULL);
+
+    xocl_drvinst_free(hdl);
+
+    return 0;
+}
+
+static int accel_deadlock_detector_probe(struct platform_device *pdev)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector;
+    struct resource *res;
+    void *priv;
+    int err = 0;
+
+    accel_deadlock_detector = xocl_drvinst_alloc(&pdev->dev, sizeof(struct xocl_accel_deadlock_detector));
+    if (!accel_deadlock_detector) {
+        err = -ENOMEM;
+        xocl_err(&pdev->dev, "xocl_drvinst_alloc failed for accel_deadlock_detector_probe , err (-ENOMEM): %d", err);
+        return -ENOMEM;
+    }
+
+    accel_deadlock_detector->dev = &pdev->dev;
+
+    priv = XOCL_GET_SUBDEV_PRIV(&pdev->dev);
+    if (priv)
+        memcpy(&accel_deadlock_detector->data, priv, sizeof(struct debug_ip_data));
+
+    platform_set_drvdata(pdev, accel_deadlock_detector);
+    mutex_init(&accel_deadlock_detector->lock);
+
+    res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+    if (!res) {
+        err = -ENOMEM;
+        xocl_err(&pdev->dev, "platform_get_resource failed for accel_deadlock_detector_probe , err (-ENOMEM): %d", err);
+        goto done;
```

Comment:
Hello @rozumx 

It's common to use goto statements in linux driver while error out since performance is not something we concern here.

And it also helps driver developers to have a nice and clean way to handle different kinds of errors

Reply:
Hello @houlz0507 (Lizhi),
I will defer the use of goto statements to you since this is 'c' driver code.

Reply:
'goto' is fine in driver code. It is used everywhere in linux kernel. We are using 'goto' in upstreaming code as well.